lib/xapi-fdcaps/operations.mli

```diff
@@ -0,0 +1,210 @@
+(*
+ * Copyright (C) 2023 Cloud Software Group
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+(** Statically enforce file descriptor capabilities using type parameters.
+
+ *)
+
+open Properties
+
+(** a file descriptor with properties
+  Upper bounds are avoided here so that this type can be used in functors
+ *)
+type +!'a props constraint 'a = (_, _) Properties.props
+
+(** like {!type:props} but with upper bounds on properties *)
+type +!'a t = 'a props constraint 'a = (_, _) Properties.t
+
+(** convenience type for declaring properties *)
+type (+!'a, +!'b) make = ('a, 'b) Properties.t t
+
+val pp : _ t Fmt.t
+(** [pp formatter t] pretty prints [t] on [formatter]. *)
+
+val dump : _ t Fmt.t
+(** [dump formatter t] prints a debug representation of [t] on [formatter]. *)
+
+val close : _ t -> unit
+(** [close t] closes t. Doesn't raise an exception if it is already closed.
+  Other errors from the underlying {!val:Unix.close} are propagated.
+ *)
+
+val with_fd : 'a t -> ('a t -> 'b) -> 'b
+(** [with_fd t f] calls [f t] and always closes [t] after [f] finishes.
+  [f] can also close [t] earlier if it wants to without a double close error.
+*)
+
+val with_fd2 : 'a t * 'b t -> ('a t * 'b t -> 'c) -> 'c
+(** [with_fd2 fd1 fd2 f] calls [f fd1 fd2] and always closes [t] after [f] finishes. *)
+
+module Syntax : sig
+  val ( let@ ) : ('a -> 'b) -> 'a -> 'b
+  (** [let@ fd = with_fd t in ... use fd] *)
+end
```

Comment:
I don't this module should be exposed, seems like an implementation detail. If users want to use let@ they can easily define it

Reply:
It will be used by other modules that are not part of this PR yet.

Reply:
I've kept this as is for now. `Operations` isn't the ideal place for this `let@` definition, but we should find some shared place for it to encourage its use (it is the preferred way of using the `with_fd` and `with_fd2` functions in this module

Reply:
Should I move it to a toplevel `syntax.mli` in `xapi-fdcaps`?